src/mol-plugin-state/manager/focus-camera/focus-first-residue.ts

```diff
@@ -0,0 +1,149 @@
+                sumZ += firstChainPositions[i + 2];
+            }
+            const averagePosition = Vec3.zero();
+            averagePosition[0] = sumX / atomIndices.length;
+            averagePosition[1] = sumY / atomIndices.length;
+            averagePosition[2] = sumZ / atomIndices.length;
+            return averagePosition;
+        } else {
+            return Vec3.create(caPositions[0], caPositions[1], caPositions[2]);
+        }
+    }
+
+}
+
+export function pcaFocus(plugin: PluginContext, options: Partial<CameraFocusOptions> & { principalAxes?: PrincipalAxes, positionToFlip?: Vec3 }) {
+    if (options?.principalAxes) {
+        const { origin, dirA, dirB, dirC } = options.principalAxes.boxAxes;
+        let toFlip: ('aroundX' | 'aroundY')[] = [];
+        if (options.positionToFlip) {
+            toFlip = getAxesToFlip(options.positionToFlip, origin, dirA, dirB);
+        }
+        toFlip.forEach((axis)=>{
+            if (axis === 'aroundY') {
+                Vec3.negate(dirC, dirC);
+            } else if (axis === 'aroundX') {
+                Vec3.negate(dirA, dirA);
+                Vec3.negate(dirC, dirC);
+            }
+        });
+        if (plugin.canvas3d) {
+            const position = Vec3();
+            Vec3.scaleAndAdd(position, position, origin, 100);
+            plugin.canvas3d.camera.setState({ position }, 0);
+            const deltaDistance = Vec3();
+            Vec3.negate(deltaDistance, position);
+            if (Vec3.dot(deltaDistance, dirC) <= 0) {
+                Vec3.negate(plugin.canvas3d.camera.position, position);
+            }
+            const up = Vec3.create(0, 1, 0);
+            if (Vec3.dot(up, dirA) <= 0) {
+                Vec3.negate(plugin.canvas3d?.camera.up, plugin.canvas3d.camera.up);
+            }
+        }
+        return { origin, dirA, dirB, dirC };
+    }
+    return {
+        origin: Vec3.zero(),
+        dirA: Vec3.zero(),
+        dirB: Vec3.zero(),
+        dirC: Vec3.zero()
+    };
+}
+
+export function getPcaTransform(group: StructureComponentRef[]): { principalAxes?: PrincipalAxes, positionToFlip?: Vec3 } | undefined {
+    const polymerStructure = group[0].cell.obj?.data;
+    if (!polymerStructure) {
+        return undefined;
+    }
+    if ('_pcaTransformData' in polymerStructure.currentPropertyData) {
+        return deepClone(polymerStructure.currentPropertyData._pcaTransformData);
+    }
+    if (!polymerStructure.units[0]?.props.polymerElements?.length) {
+        polymerStructure.currentPropertyData._pcaTransformData = undefined;
+        return undefined;
+    }
+    const positions = getPolymerPositions(polymerStructure);
+    const positionToFlip = getFirstResidueOrAveragePosition(polymerStructure, positions);
+    polymerStructure.currentPropertyData._pcaTransformData = { principalAxes: PrincipalAxes.ofPositions(positions), positionToFlip };
+    return deepClone(polymerStructure.currentPropertyData._pcaTransformData);
```

Comment:
Does this need a deep clone or is a shallow clone actually enough?

Reply:
Shallow clone will cause the view to flip after each click in some structure.